File: 프로그래머스/2/84512. 모음 사전/모음 사전.py
"""
    Make dictionary
"""
# from itertools import product

# def solution(word):
#     dictionary = [''.join(p) for p in product(['', 'A', 'E', 'I', 'O', 'U'], repeat = 5)]
#     dictionary = sorted(list(set(dictionary)))
#     return dictionary.index(word)

"""
    Use Recursion
"""
def dict_maker(word, length, dictionary):
    if length == 6: return
    if word != '': dictionary.append(word)
    for alphabet in ['A', 'E', 'I', 'O', 'U']:
        dict_maker(''.join([word, alphabet]), length + 1, dictionary)

def solution(word):
    dictionary = []
    dict_maker('', 0, dictionary)
    return dictionary.index(word) + 1

File: 프로그래머스/2/84512. 모음 사전/모음 사전.js
/**
 * Use Recursion
 */
const dict_maker = (word, len, dict) => {
    if (len === 6) return;
    if (word !== '') dict.push(word);
    ['A', 'E', 'I', 'O', 'U'].forEach(alphabet => dict_maker(word.concat(alphabet), len + 1, dict));
}

function solution(word) {
    const dictionary = [];
    dict_maker('', 0, dictionary)
    return dictionary.indexOf(word) + 1;
}